File: formatters/mla9.py
"""
MLA 9th Edition formatter.

Layout:
  - No separate title page.
  - First page starts with a left-aligned block (student name / instructor / course / date),
    then a centered plain-text title, then the essay body immediately follows.
  - Page numbers: "LastName #" in top-right header on every page.
  - Works Cited page: "Works Cited" centered (not bold), hanging indent, double-spaced.
"""

from docx import Document
from docx.enum.text import WD_ALIGN_PARAGRAPH, WD_LINE_SPACING
from docx.shared import Inches, Pt

from models import EssayJSON
from .base import (
    add_body,
    apply_document_defaults,
    add_page_numbers,
    add_references_page,
    doc_to_bytes,
    extract_last_name,
    set_run_font,
    setup_section,
)


def build(essay: EssayJSON, config: dict) -> bytes:
    doc = Document()
    apply_document_defaults(doc, config)
    section = doc.sections[0]
    setup_section(section, config)

    last_name = extract_last_name(essay.author_name_placeholder)
    add_page_numbers(section, config, last_name=last_name)  # "LastName #" on every page
    _first_page_block(doc, essay, config)
    add_body(doc, essay.body_markdown, config, _heading)
    doc.add_page_break()
    add_references_page(doc, essay.references, config)

    return doc_to_bytes(doc)


# ── First-page header block ──────────────────────────────────────────────────

def _first_page_block(doc: Document, essay: EssayJSON, config: dict) -> None:
    # Top-left four-line block: name, instructor, course, date
    for text in [
        essay.author_name_placeholder,
        essay.instructor_placeholder,
        essay.course_placeholder,
        essay.date,
    ]:
        p = doc.add_paragraph()
        p.alignment = WD_ALIGN_PARAGRAPH.LEFT
        p.paragraph_format.line_spacing_rule = WD_LINE_SPACING.DOUBLE
        p.paragraph_format.space_before = Pt(0)
        p.paragraph_format.space_after = Pt(0)
        p.paragraph_format.first_line_indent = Inches(0)
        run = p.add_run(text)
        set_run_font(run, config)

    # Title: centered, plain text — no bold, no italic, no quotes
    title_para = doc.add_paragraph()
    title_para.alignment = WD_ALIGN_PARAGRAPH.CENTER
    title_para.paragraph_format.line_spacing_rule = WD_LINE_SPACING.DOUBLE
    title_para.paragraph_format.space_before = Pt(0)
    title_para.paragraph_format.space_after = Pt(0)
    title_para.paragraph_format.first_line_indent = Inches(0)
    run = title_para.add_run(essay.title)
    set_run_font(run, config)


# ── Headings ─────────────────────────────────────────────────────────────────

def _heading(doc: Document, text: str, level: int, config: dict) -> None:
    """MLA does not mandate a specific heading style — left-aligned, plain text."""
    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.LEFT
    p.paragraph_format.line_spacing_rule = WD_LINE_SPACING.DOUBLE
    p.paragraph_format.space_before = Pt(0)
    p.paragraph_format.space_after = Pt(0)
    p.paragraph_format.first_line_indent = Inches(0)
    run = p.add_run(text)
    set_run_font(run, config)

File: src/Formatters/Mla9.php
<?php

namespace App\Formatters;

use App\Models\EssayJSON;
use PhpOffice\PhpWord\Element\Section;
use PhpOffice\PhpWord\PhpWord;
use PhpOffice\PhpWord\Shared\Converter;

/**
 * MLA 9th Edition formatter.
 *
 * Layout:
 *   - No separate title page.
 *   - First page starts with a left-aligned block (student name / instructor / course / date),
 *     then a centered plain-text title, then the essay body immediately follows.
 *   - Page numbers: "LastName #" in top-right header on every page.
 *   - Headings: Level 1 (##) bold flush-left, Level 2 (###) italic flush-left.
 *   - Works Cited page: "Works Cited" centered (not bold), hanging indent, double-spaced.
 */
class Mla9
{
    public static function build(EssayJSON $essay, array $config): string
    {
        $phpWord = new PhpWord();
        Base::applyDocumentDefaults($phpWord, $config);

        $section = $phpWord->addSection();
        Base::setupSection($section, $config);

        $lastName = Base::extractLastName($essay->authorNamePlaceholder);
        Base::addPageNumbers($section, $config, $lastName);
        self::firstPageBlock($section, $essay, $config);
        Base::addBody($section, $essay->bodyMarkdown, $config, [self::class, 'heading']);
        $section->addPageBreak();
        Base::addReferencesPage($section, $essay->references, $config);

        return Base::docToBytes($phpWord);
    }

    // ── First-page header block ───────────────────────────────────────────────

    private static function firstPageBlock(Section $section, EssayJSON $essay, array $config): void
    {
        // Top-left four-line block: name, instructor, course, date
        foreach ([
            $essay->authorNamePlaceholder,
            $essay->instructorPlaceholder,
            $essay->coursePlaceholder,
            $essay->date,
        ] as $text) {
            $section->addText($text, Base::fontStyle($config), [
                'alignment'   => 'left',
                'lineHeight'  => 2.0,
                'spaceBefore' => 0,
                'spaceAfter'  => 0,
                'indentation' => ['firstLine' => 0],
            ]);
        }

        // Title: centered, plain text — no bold, no italic
        $section->addText($essay->title, Base::fontStyle($config), [
            'alignment'   => 'center',
            'lineHeight'  => 2.0,
            'spaceBefore' => 0,
            'spaceAfter'  => 0,
            'indentation' => ['firstLine' => 0],
        ]);
    }

    // ── Headings ──────────────────────────────────────────────────────────────

    public static function heading(Section $section, string $text, int $level, array $config): void
    {
        if ($level === 0) {
            Base::addDocumentTitle($section, $text, $config, false); // MLA: title NOT bold
            return;
        }

        $base = [
            'lineHeight'  => 2.0,
            'spaceBefore' => 0,
            'spaceAfter'  => 0,
            'indentation' => ['firstLine' => 0],
        ];

        // MLA heading levels (per MLA Handbook):
        // L1 ##:    bold, flush left
        // L2 ###:   italic, flush left
        // L3 ####:  centered, bold
        // L4 #####: centered, italic
        // L5 ######: underlined, flush left
        $specs = [
            1 => ['alignment' => 'left',   'bold' => true,  'italic' => false, 'underline' => false],
            2 => ['alignment' => 'left',   'bold' => false, 'italic' => true,  'underline' => false],
            3 => ['alignment' => 'center', 'bold' => true,  'italic' => false, 'underline' => false],
            4 => ['alignment' => 'center', 'bold' => false, 'italic' => true,  'underline' => false],
            5 => ['alignment' => 'left',   'bold' => false, 'italic' => false, 'underline' => true],
        ];
        $s = $specs[$level] ?? $specs[5];

        $textRun = $section->addTextRun(array_merge($base, ['alignment' => $s['alignment']]));
        $textRun->addText($text, [
            'name'      => $config['font_family'],
            'size'      => $config['font_size'],
            'bold'      => $s['bold'],
            'italic'    => $s['italic'],
            'underline' => $s['underline'] ? 'single' : 'none',
        ]);
    }
}
